### ISSUE TITLE
TorchModel move(value) recursive calls

### ISSUE THREAD

**Issue by AlexWan0** [HLR/DomiKnowS#290]
For a `data_item` many non-tensor nested lists, the `move` function recursively calls itself many times, with each call taking a long time for models with large numbers of parameters (e.g. ones with lots of learners).

[data_item used, for reference](https://pastebin.com/3MLfvb1G)

The `move` function calls itself 70615 times with this `data_item`. This number of calls is made regardless of the number of learners there are, but increasing the number of learners increases the time it takes to call
`parameters = list(self.parameters())`
which is used to find the device to move the tensor to.
Each call of this line takes 4.220008850097656e-05 for 10 learners and 0.005646944046020508 for 1076 learners.

Since the only purpose of this call is to find the correct device to use, it may be better to define the device once, on initialization of the `TorchModel`, then use that value in subsequent calls.

i.e. adding
```
parameters = list(self.parameters())
self.device = next(self.parameters()).device
```

to `__init__ `in `TorchModel` and

using `self.device`, instead of redefining it every time `move` is called.

Overall, the `move` call took 394 seconds, and one entire train iteration took 397 seconds (with 1076 learners). With the fix, move took 0.10 seconds, and a training iteration took 3.30 seconds.

**Comment by hfaghihi15** [comment]
@guoquan Do you have any idea about this issue? ( I am asking this because you implemented the device features, just want to make sure that there is nothing here that we are missing from your perspective)

**Comment by AlexWan0** [comment]
Alright, I will do that.

**Comment by AlexWan0** [comment]
Yeah, the fix works. I've changed my code slightly since then, so the entire move operation takes 240.6984 seconds total without the fix, and 0.869 seconds with the fix.
